server/handlers/product.js

```diff
@@ -1,80 +1,262 @@
+    try {
+        const user = await db.User.findById(req.params.id);
+        if (!user) {
+            return res.status(401).json({ error: "User do not exist" });
+        }
+        const cart = user.cart;
+        return res.status(200).json(cart);
+    } catch (err) {
+        return next(err);
+    }
+};
+
+const addProductToCart = async (req, res, next) => {
+    try {
+        // find the user
+        const user = await db.User.findById(req.params.id);
+        if (!user) {
+            return res.status(401).json({ error: "User do not exist" });
+        }
+
+        // find the product
+        const quantity = req.body.quantity;
+        const product = await db.Product.findById(req.body.productId);
+        if (!product || !product.stockNum) {
+            return res.status(401).json({ error: "Product not found" });
+        } else if (product.stockNum < quantity) {
+            return res
+                .status(401)
+                .json({ error: "Insufficient stock of this product" });
+        }
+
+        // add product into cart
+        if (user.cart.has(product.id)) {
+            const currQuantity = user.cart.get(product.id);
+            user.cart.set(product.id, currQuantity + quantity);
+        } else user.cart.set(product.id, quantity);
+
+        // save user
+        await user.save();
+
+        const { id, username, category, avatarUrl, cart } = user;
+        return res.status(200).json({
+            id,
+            username,
+            category,
+            avatarUrl,
+            cart,
+        });
+    } catch (err) {
+        return next(err);
+    }
+};
+
+const getProductByIdInCart = async (req, res, next) => {
+    try {
+        const user = await db.User.findById(req.params.id);
+        if (!user) {
+            return res.status(401).json({ error: "User do not exist" });
+        }
+
+        if (user.cart.has(req.params.productId)) {
+            const product = await db.Product.findById(req.params.productId);
+            if (!product || !product.stockNum) {
+                return res
+                    .status(401)
+                    .json({ error: "No such product in stock" });
+            }
+
+            return res.status(200).json({
+                product: product.name,
+                quantity: user.cart.get(req.params.productId),
+            });
+        } else {
+            return res.status(201).json({ error: "No such product in cart" });
+        }
+    } catch (err) {
+        return next(err);
+    }
+};
+
+const updateProductQuantityInCart = async function (req, res, next) {
+    try {
+        const user = await db.User.findById(req.params.id);
+        if (!user) {
+            return res.status(401).json({ error: "User do not exist" });
+        }
+
+        // lock cart for multi fast update
+        const cartLock = await acquireCartLock(req.params.id);
+
+        if (user.cart.has(req.params.productId)) {
+            // const quantity =
```

Comment:
why not use 'const quantity = user.cart.get(req.params.productId) + req.body.quantity;'?

Reply:
It depends on the React's method we use to control the quantity of the product in the cart. Both  `const quantity = req.body.quantity;` and `const quantity = user.cart.get(req.params.productId) + req.body.quantity;` are available. We will decide it later in React design.

Reply:
> why not use 'const quantity = user.cart.get(req.params.productId) + req.body.quantity;'?

It depends on the Frontend methods we use to control the quantity of the product in the cart. Both `const quantity = req.body.quantity;` and `const quantity = user.cart.get(req.params.productId) + req.body.quantity;` are available. We will decide later in React design.